### DISCUSSION TITLE
Carousel re-init on very small width change

### DISCUSSION THREAD

**Discussion by hamidrezahanafi** [davidjerleke/embla-carousel#569] [Help]
Hello everyone,
I checked that resize observer can react to very small width/height changes for example in my case the change was from 163.890625 => 163.89060974121094 (which I think it's not actual change and is rounding of the browser)
Is it a good idea if the condition of new size vs old size to be more than 1 pixel instead of not equal?

**Comment by davidjerleke** [comment]
Hi @hamidrezahanafi,

Thanks for your input. Yes it will **react** to **any change** because that's how ResizeObserver works. Sometimes this may be the wanted behaviour and sometimes not. I can't rule out that this is the wanted behaviour in some cases. It's impossible for me to know.

However, if you [read the docs](https://www.embla-carousel.com/api/options/#watchresize), Embla gives you the freedom to provide your own callback and customise the resize behaviour. For convenience, it gives you the `emblaApi` and resize observer `entries`:

```ts
const [emblaRef, emblaApi] = useEmblaCarousel({
  watchResize: (emblaApi, resizeObserverEntries) => {

    // you can also access the current rects like this:
    const { slideRects, containerRect } = emblaApi.internalEngine()

    // do your own thing and run reInit() when you think the carousel should reset and pick up changes
    if (yourCustomConditionIsMet) emblaApi.reInit()
    
    // return false so Embla doesn't fire it's default resize behaviour after your custom callback
    return false
  }
})
```

Best,
David

**Reply by hamidrezahanafi** [reply]
Thanks for your reply @davidjerleke , yeah I know and already using that.

I think I should have explained it better. I do think it's an actual library issue.

The trigger to resize observer is coming from a height change in the slides but we only check the widths for an axis x slider. In this case width didn't change but upon checking the old width `163.890625` and new width `163.89060974121094` it seems that style width and observer width have different precision.  `163.890625` is the actual rounded number of `163.89060974121094`.
So library trigger a resize event without any width change which I think is library issue.

**Reply by davidjerleke** [reply]
@hamidrezahanafi thanks for explaining further.

I see. Well browsers can't render super sub-pixels anyway som maybe we should do something like this [here](https://github.com/davidjerleke/embla-carousel/blob/master/packages/embla-carousel/src/components/ResizeHandler.ts#L48) instead:

```ts
if (mathAbs(newSize - lastSize) >= 0.5) {
  // ...trigger resize
}
```

...which means only do something when the container or any slide size changes by half a pixel?

Best,
David

**Reply by hamidrezahanafi** [reply]
Yes that's exactly what I had in mind, I think it would work.

**Reply by davidjerleke** [reply]
@hamidrezahanafi feel free to create a PR if you have time. Import `mathAbs` from utils. Maybe we can do `0.2px` instead of `0.5px`. I guess that would still solve the problem.

Best,
David

**Reply by hamidrezahanafi** [reply]
https://github.com/davidjerleke/embla-carousel/pull/570

**Reply by davidjerleke** [reply]
Thanks a ton @hamidrezahanafi ⭐!

**Comment by davidjerleke** [comment]
- Tracked in #571.
